File: 6.Python/sentimental-credit/credit.py
# Prompt user for card number
# Check card number using Luhn algorithm
# Print results
def main():
    card_number = input("Card number: ")
    if luhn(card_number) == False:
        print("INVALID")
    else:
        print(card_type(card_number))


# Use list comprehension to split the card number in two
# Double every second digit from the right (even digits)
# Returns True if == 0, or False if != 0
def luhn(card):
    even = [int(i) * 2 if int(i) * 2 <= 9 else int(i) * 2 - 9 for i in card[-2::-2]]
    odd = [int(j) for j in card[-1::-2]]
    return (sum(even + odd) % 10) == 0


# Check length of card, and starting digits
# Use a list comprehension within the any function
# An any function returns True if at least 1 element in the iterable is true
# Return results as a string to be printed in main
def card_type(card):
    if len(card) in (13, 16) and card.startswith("4"):
        return "VISA"
    elif len(card) == 16 and any(card.startswith(str(i)) for i in range(51, 56)):
        return "MASTERCARD"
    elif len(card) == 15 and card.startswith(("34", "37")):
        return "AMEX"
    else:
        return "INVALID"


main()

File: 1.C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool check_luhn(long card);
void report_type(bool luhn, long card);

int main(void)
{
    // Prompt user for card number
    long card;
    do
    {
        card = get_long("Card Number: ");
    }
    while (card < 1);

    // Check luhn
    bool luhn = check_luhn(card);

    // Report card type
    report_type(luhn, card);
}


bool check_luhn(long card)
{
    bool every_other_digit = 0;
    int checksum = 0;
    int i, n, digit;
    do
    {
        // Sum up digits
        if (!every_other_digit)
        {
            digit = (card % 10);
        }
        // Multiply every other digit by 2 and sum up digits
        else
        {
            n = card % 10 * 2;
            digit = (n > 9) ? n - 9 : n;
        }
        checksum += digit;
        card /= 10;
        every_other_digit = !every_other_digit;
    } while (card > 0);
    // Return True if checksum = 0
    bool luhn = checksum % 10 == 0;
    return luhn;
}


void report_type(bool luhn, long card)
{
    // Grab card prefix
    long prefix = card;
    do
    {
        prefix /= 10;
    } while (prefix > 100);
    
    // Calculate card length
    int length = 0;
    do
    {
        card /= 10;
        length++;
    } while (card > 0);
    
    // Print final result
    string type = "INVALID\n";
    if (luhn)
    {
        if (length == 16 && prefix > 50 && prefix < 56)
        {
            type = "MASTERCARD\n";
        }
        else if ((length == 13 || length == 16) && prefix /10 == 4)
        {
            type = "VISA\n";
        }
        else if (length == 15 && (prefix == 34 || prefix == 37))
        {
            type = "AMEX\n";
        }    
    }
    printf("%s", type);
}
